src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php

```diff
@@ -177,12 +182,42 @@ public function validate(mixed $entity, Constraint $constraint)
          * which is the same as the entity being validated, the criteria is
          * unique.
          */
-        if (!$result || (1 === \count($result) && current($result) === $entity)) {
+        if (!$result || (1 === \count($result) && current($result) === $value)) {
             return;
         }
 
-        $errorPath = $constraint->errorPath ?? $fields[0];
-        $invalidValue = $criteria[$errorPath] ?? $criteria[$fields[0]];
+        /* If a single entity matched the query criteria, which is the same as
+         * the entity being updated by validated object, the criteria is unique.
+         */
+        if (!$isEntity && !empty($constraint->identifierFieldNames) && 1 === \count($result)) {
+            if (!\is_array($constraint->identifierFieldNames) && !\is_string($constraint->identifierFieldNames)) {
+                throw new UnexpectedTypeException($constraint->identifierFieldNames, 'array');
+            }
+
+            $identifierFieldNames = (array) $constraint->identifierFieldNames;
+
+            $fieldValues = $this->getFieldValues($value, $class, $identifierFieldNames);
+            if (array_values($class->getIdentifierFieldNames()) != array_values($identifierFieldNames)) {
+                throw new ConstraintDefinitionException(sprintf('The "%s" entity identifier field names should be "%s", not "%s".', $entityClass, implode(', ', $class->getIdentifierFieldNames()), implode(', ', $constraint->identifierFieldNames)));
+            }
+
+            $entityMatched = true;
+
+            foreach ($identifierFieldNames as $identifierFieldName) {
+                $propertyValue = $this->getPropertyValue($entityClass, $identifierFieldName, current($result));
+                if ($fieldValues[$identifierFieldName] !== $propertyValue) {
+                    $entityMatched = false;
+                    break;
+                }
+            }
+
+            if ($entityMatched) {
+                return;
+            }
+        }
+
+        $errorPath = $constraint->errorPath ?? current($fields);
+        $invalidValue = $criteria[$errorPath] ?? $criteria[current($fields)];
```

Comment:
Why this change?

Reply:
Fields may now be a key=>value array (Example 2 from the description of PR), so there might be no item with key 0.
We still need the first item of the array.
'testValidateMappingOfFieldNames' will fail without this change.